feat: explicit layout apply via HangingProtocolService

sedghi opened the pull request:
### PR Checklist

- [ ] Brief description of changes
- [ ] Links to any relevant issues
- [ ] Required status checks are passing
- [ ] User cases if changes impact the user's experience
- [ ] `@mention` a maintainer to request a review

<!--
  Links
  -->

<!-- prettier-ignore-start -->
[blog]: https://circleci.com/blog/triggering-trusted-ci-jobs-on-untrusted-forks/
[script]: https://github.com/jklukas/git-push-fork-to-upstream-branch
<!-- prettier-ignore-end -->

wayfarer3130 commented on extensions/cornerstone/src/commandsModule.ts:
```diff
@@ -213,6 +218,8 @@ const commandsModule = ({ servicesManager }) => {
           },
         });
+
+        return;
```
This seems an odd thing to have a return statement here - it returns at the end of the method regardless, and there isn't more code after that.

wayfarer3130 commented on extensions/cornerstone/src/commandsModule.ts:
```diff
@@ -415,6 +422,122 @@ const commandsModule = ({ servicesManager }) => {
+      });
+
+      const hpViewports: Types.HangingProtocol.Viewport[] = [
+        'axial',
```
Can you add acquisition here?

wayfarer3130 replied:
I see what you are doing, you are generating a hanging protocol to apply.  Is there any reason not to allow applying a generic hanging protocol, and just define it in the module in the normal way?  That way one could define buttons/commands to apply arbitrary hanging protocols, and you wouldn't have this large chunk of data generation in the middle of the commands module.

wayfarer3130 commented on extensions/cornerstone/src/commandsModule.ts:
```diff
@@ -415,6 +422,122 @@ const commandsModule = ({ servicesManager }) => {
+        };
+      });
+
+      const mprProtocol: Types.HangingProtocol.Protocol = {
```
Please import this from it's own file

wayfarer3130 commented on modes/longitudinal/src/toolbarButtons.js:
```diff
@@ -286,6 +286,23 @@ const toolbarButtons = [
+  //     commands: [
+  //       {
+  //         commandName: 'activateMPRForActiveViewport',
+  //         commandOptions: {},
```
I suggest defining a module exported hanging protocol name here that can be applied.  That way, all hanging protocols are defined in the same way, and a mode can changes the "MPR" layout to whatever they want just by switching the HP to apply.    I would also suggest specifying the default stage to start with, at least as a possible option.

wayfarer3130 commented on platform/core/src/types/HangingProtocol.ts:
```diff
@@ -0,0 +1,134 @@
+type DisplaySetInfo = {
```
Thank you for adding the typing for this - it really helps to get it typed.

wayfarer3130 commented on platform/core/src/types/index.ts:
```diff
@@ -0,0 +1,5 @@
+import type * as HangingProtocol from './HangingProtocol'
+
+export {
```
Why isn't this one export type { ...
It looks like it is otherwise the same as the import.

wayfarer3130 commented on extensions/cornerstone/src/commandsModule.ts:
```diff
@@ -415,6 +422,122 @@ const commandsModule = ({ servicesManager }) => {
         (activeViewportIndex - 1 + viewports.length) % viewports.length;
       ViewportGridService.setActiveViewportIndex(nextViewportIndex);
     },
+    activateMPRForActiveViewport: () => {
```
Can you call this applyHangingProtocol, taking the name of a HP extension export as the parameter and optionally a stage.

wayfarer3130 commented on platform/core/src/services/HangingProtocolService/HangingProtocolService.ts:
```diff
@@ -392,7 +401,7 @@ class HangingProtocolService {
 
     this.customImageLoadPerformed = false;
-    const { type: layoutType } = stageModel.viewportStructure;
+    const { layoutType } = stageModel.viewportStructure;
```
Is it necessary to rename the field name?  Probably it is, given the use of type as a keyword.

sedghi replied:
Yes it was actually a bug, it is defined as layoutType in the protocols

wayfarer3130 commented on platform/core/src/services/HangingProtocolService/HangingProtocolService.ts:
```diff
@@ -313,7 +297,20 @@ class HangingProtocolService {
 
-  _setProtocol(protocol) {
+  public applyProtocol(
+    protocol: HangingProtocol.Protocol,
```
It should be possible to apply the protocol by name instead of handing it a full protocol in order to re-apply existing protocols, eg to go back to the previously applied protocol (un-apply the MPR mode).

wayfarer3130 commented on platform/core/src/services/HangingProtocolService/HangingProtocolService.ts:
```diff
@@ -313,7 +297,20 @@ class HangingProtocolService {
+    matchingDisplaySets: Record<string, HangingProtocol.DisplaySetMatchDetails>
+  ): void {
+    protocol = this._validateProtocol(protocol);
+    this.protocols.push(protocol);
```
It also changes the initial display conditions if you add it to the protocols list - what happens if you:
1. Display a study in the default protocol
2. Apply the MPR mode
3. Apply the "default" protocol again as a apply most applicable protocol - except this time it hits the MPR as the most applicable protocol.

I'd rather see the protocol name be the actual id of a protocol here to apply - the protocols can then be added in the default fashion, and to turn it off, just avoid setting any conditions on the protocol.   That also allows your MPR check to be generic, just call the study level protocol conditions.  Have the extension register a custom attribute that checks if the study contains a display set for which the response for the check is true.

wayfarer3130 commented on platform/core/src/services/HangingProtocolService/HangingProtocolService.ts:
```diff
@@ -313,7 +297,20 @@ class HangingProtocolService {
-  _setProtocol(protocol) {
+  public applyProtocol(
+    protocol: HangingProtocol.Protocol,
+    matchingDisplaySets: Record<string, HangingProtocol.DisplaySetMatchDetails>
```
It should be allowed to supply the matchingDisplaySets optionally - it is entirely possible to just reapply a specific protocol without specifying the display sets

sedghi commented:
@wayfarer3130 

1. Changed the HP Service to be able to register both Object and Protocol Generator (dynamic) protocols via `id` 
2. Changed the protocol registration to happen at the extension registration, previously it was done by the modes which was wrong, now a mode can define which protocol it should run with, if undefined we will run protocol Matching rules. 
3. Recent PR added inheritance to ViewportGrid which should not happen, as we cannot assume the next displaySets and previous displaySets on the same viewport share the same options.

sedghi commented:
I need to change docs as well, after we settle on the changes @sedghi

wayfarer3130 commented on extensions/default/src/index.js:
```diff
@@ -2,9 +2,9 @@ import getDataSourcesModule from './getDataSourcesModule.js';
 import getLayoutTemplateModule from './getLayoutTemplateModule.js';
 import getPanelModule from './getPanelModule';
 import getSopClassHandlerModule from './getSopClassHandlerModule.js';
-import getHangingProtocolModule from './getHangingProtocolModule.js';
```
Is there a good reason for this change here?  It seems an empty change...

sedghi replied:
That was a mistake

wayfarer3130 commented on extensions/tmtv/src/getHangingProtocolModule.js:
```diff
@@ -15,6 +15,17 @@ const ptCT = {
+      weight: 1,
+      attribute: 'ModalitiesInStudy',
+      constraint: {
+        contains: {
```
Use the contains: ['PT', 'CT'] syntax - it is a little easier to read.

wayfarer3130 commented on platform/core/src/services/HangingProtocolService/HangingProtocolService.ts:
```diff
@@ -116,30 +94,91 @@ class HangingProtocolService {
-      if (this.protocols.indexOf(protocol) === -1) {
-        this.protocols.push(this._validateProtocol(protocol));
+  /**
+   * Returns the protocol with the given id
```
Please add documentation that it will call the generator function to create/update the protocol as necessary.

wayfarer3130 commented on platform/core/src/services/HangingProtocolService/HangingProtocolService.ts:
```diff
@@ -524,6 +657,7 @@ class HangingProtocolService {
         const seriesMatchDetails = this.protocolEngine.findMatch(
           displaySet,
           seriesMatchingRules,
+          // Todo: why we have images here since the matching type does not have it
```
The reason to have images here is to allow getting the metadata to read attributes not included in the study by default - there are other possible patient/study level attributes.

wayfarer3130 commented:
LGTM after the docs and couple of small changes are applied.
